File: Easy/28_Find_the_Index_of_the_First_Occurrence_in_a_String/Solution.py
from typing import List, Optional


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        haystack_length = len(haystack)
        needle_length = len(needle)

        if needle_length == 0:
            return 0
        if haystack_length < needle_length:
            return -1

        for i in range(haystack_length - needle_length + 1):
            j = 0
            while j < needle_length and haystack[i + j] == needle[j]:
                j += 1
            if j == needle_length:
                return i

        return -1


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    print(solution.strStr("sadbutsad", "sad"))  # Output: 0

    # test cases 2
    print(solution.strStr("leetcode", "leeto"))  # Output: -1

File: Easy/28_Find_the_Index_of_the_First_Occurrence_in_a_String/Solution.java



public class Solution {
    public int strStr(String haystack, String needle) {
        int index = -1;
        int haystackLength = haystack.length();
        int needleLength = needle.length();

        char[] haystackArr = haystack.toCharArray();
        char[] needleArr = needle.toCharArray();

        if (needleLength == 0) {
            return 0;
        }
        if (haystackLength < needleLength) {
            return -1;
        }

        for (int i = 0; i <= haystackLength - needleLength; i++) {
            int j;

            for (j = 0; j < needleLength; j++) {
                if (haystackArr[i + j] != needleArr[j]) {
                    break;
                }
            }

            if (j == needleLength) {
                index = i;
                break;
            }
        }
        
        return index;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String haystack1 = "sadbutsad";
        String needle1 = "sad";
        System.out.println(solution.strStr(haystack1, needle1)); // Output:

        // test cases 2
        String haystack2 = "leetcode";
        String needle2 = "leeto";
        System.out.println(solution.strStr(haystack2, needle2)); // Output: -1
        
    }
}